### PR TITLE
feat: implement push ubi image by explicitly running commands for ubi and centos

### PR DESCRIPTION
**Pull request by Sn0rt**
1. build the APISIX, the APISIX base with the UBI image.
2. update ci to push the ubi.rpm package

### CODE REVIEW

File: utils/publish-rpm.sh [hunk 1]
```diff
@@ -149,41 +141,44 @@ init_cos_utils)
 repo_init)
     # create basic repo directory structure
     # useful when a new repo added
-    func_repo_init /tmp "$dist"
+    func_repo_init /tmp
     ;;
 repo_backup)
-    func_repo_backup "${VAR_COS_BUCKET_REPO}" "$dist" "${TAG_DATE}"
+    func_repo_backup "${VAR_COS_BUCKET_REPO}" "centos" "${TAG_DATE}"
+    func_repo_backup "${VAR_COS_BUCKET_REPO}" "centos" "${TAG_DATE}"
```
Comment (tao12345666333) [inline utils/publish-rpm.sh:148]: redhat

File: .github/workflows/package-apisix-dashboard-rpm-ubi.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
+name: package apisix-dashabord rpm for ubi
```
Comment (Revolyssup) [inline .github/workflows/package-apisix-dashboard-rpm-ubi.yml:1]: @Sn0rt The ubi package is not required for dashboard as said by @tao12345666333

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -52,29 +49,25 @@ jobs:
...
 
       - name: Build apisix RPM Package
-        if: ${{ startsWith(steps.tag_type.outputs.version, 'apisix/') }}
+        if: ${{ startsWith(steps.tag_type.outputs.version, 'apisix/') || startsWith(steps.tag_type.outputs.version, 'build-all/') }}
         env:
-          APISIX_TAG_VERSION: ${{ steps.tag_env.outputs.version}}
+          APISIX_TAG_VERSION: 3.3.0
         run: |
           # build apisix
-          make package type=rpm app=apisix openresty=apisix-base checkout=${APISIX_TAG_VERSION} version=${APISIX_TAG_VERSION} image_base=centos image_tag=7
```
Comment (Revolyssup) [inline .github/workflows/publish.yml:60]: Why does this workflow remove building for el7 and el8?

**Review by Revolyssup — CHANGES_REQUESTED** [review]

File: .github/workflows/publish.yml (hunk 3 above)
Comment (Sn0rt) [reply .github/workflows/publish.yml:60]: it's just POC, for quick build a rpm of redhat

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -52,9 +49,9 @@ jobs:
           echo "##[set-output name=version;]$(echo ${GITHUB_REF#refs/tags/})"
 
       - name: Build apisix RPM Package
-        if: ${{ startsWith(steps.tag_type.outputs.version, 'apisix/') }}
+        if: ${{ startsWith(steps.tag_type.outputs.version, 'apisix/') || startsWith(steps.tag_type.outputs.version, 'build-all/') }}
```
Comment (tao12345666333) [inline .github/workflows/publish.yml:52]: Why you add this condition?

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -64,9 +61,9 @@ jobs:
           echo "TARGET_APP=apisix" >> "$GITHUB_ENV"
 
       - name: Build apisix-base RPM Package
-        if: ${{ startsWith(steps.tag_type.outputs.version, 'apisix-base/') }}
+        if: ${{ startsWith(steps.tag_type.outputs.version, 'apisix-base/') || startsWith(steps.tag_type.outputs.version, 'build-all/') }}
```
Comment (tao12345666333) [inline .github/workflows/publish.yml:64]: Ditto

File: .github/workflows/publish.yml (hunk 4 above)
Comment (Revolyssup) [reply .github/workflows/publish.yml]: this must have been for debugging. I removed it from everywhere. @tao12345666333

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -114,59 +130,29 @@ jobs:
...
           GPG_NAME: ${{ secrets.GPG_NAME }}
           GPG_MAIL: ${{ secrets.GPG_MAIL }}
         run: |
           echo "${{ secrets.RPM_GPG_PRIV_KEY }}" >> /tmp/rpm-gpg-publish.private
           echo "${{ secrets.RPM_GPG_PASSPHRASE }}" >> /tmp/rpm-gpg-publish.passphrase
           echo "${{ secrets.RPM_GPG_PUB_KEY }}" >> /tmp/rpm-gpg-publish.public
-          sudo -E ./utils/publish-rpm.sh rpm_gpg_sign centos
-          sudo -E ./utils/publish-rpm.sh repo_package_sync centos
-
-      - name: RPM repo backup for centos
```
Comment (Sn0rt) [inline .github/workflows/publish.yml:140]: @Revolyssup keep this code I think if not POC.

**Comment by leslie-tsang** [comment]
@Revolyssup Please take a look. :)

File: .github/workflows/publish.yml (hunk 6 above)
Comment (Revolyssup) [reply .github/workflows/publish.yml:140]: Re-added it

**Comment by Revolyssup** [comment]
> @Revolyssup Please take a look. :)

Made suggested changes

**Review by Revolyssup — APPROVED** [review]

**Comment by tao12345666333** [comment]
The current image based on ubi has been successfully built. We don't need to hardcode temporary builds anymore. Please modify this PR with the code that can be used officially. Thank you.

**Comment by Revolyssup** [comment]
> The current image based on ubi has been successfully built. We don't need to hardcode temporary builds anymore. Please modify this PR with the code that can be used officially. Thank you.

This PR doesn't hardcode anything per se, the naming of the PR might be inappropriate. Previously I had added the support to reuse same script by passing different variables to it which caused the issue here https://github.com/api7/apisix-build-tools/actions/runs/5023950452/jobs/9009125731#step:12:150
This PR, instead of reusing commands with different variables, explicitly runs twice - once with ubi8.6 and once with centos

**Comment by tao12345666333** [comment]
@Revolyssup Can you explain the code below? Obviously, they did hardcode "redhat".

https://github.com/api7/apisix-build-tools/pull/303/files#diff-a0492722e5dec6c4dc68f922ef1edf557de1d0f5e81420b2b2fecebd2fde97b9R164-R170****

https://github.com/api7/apisix-build-tools/blob/27895748507267a596ee0a54bd359572e3729b30/utils/publish-rpm.sh#L159-L170

**Comment by tao12345666333** [comment]
Once this PR is merged, does it mean that packages related to CentOS and Debian cannot be updated normally?

**Comment by Revolyssup** [comment]
> Once this PR is merged, does it mean that packages related to CentOS and Debian cannot be updated normally?

This script is only for centos and redhat. Previously it was only for centos so it had centos hardcoded in it. Now that hardcoding is for both centos and redhat. Each function is called twice, once with centos as argument and once with redhat. Thanks for pointing out above, there centos wasn't passed and only redhat was. I added centos on missing places.

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**